plugins/mac-avcapture/OBSAVCapture.h

```diff
@@ -336,6 +350,50 @@ typedef struct av_capture_info {
 
 #pragma mark - Static helper functions
 
+/// Compare two `AVCaptureDeviceFormat`s for purposes of sorting in the properties window.
+static NSComparator const OBSAVCaptureDeviceFormatCompare =
+    ^NSComparisonResult(AVCaptureDeviceFormat *a, AVCaptureDeviceFormat *b) {
+        CMVideoDimensions aDimensions = CMVideoFormatDescriptionGetDimensions(a.formatDescription);
```

Comment:
Have you tried calculating a "bandwidth" for each format (which would necessarily include resolution, number of planes, size per plane, colour depth, etc.) so you get a nice single numerical value representing a format and then you can just compare the value of `A` and `B`.

Also the default names for custom comparators in Swift is `lhs` and `rhs`, which I'd prefer here as well.

Reply:
> Have you tried calculating a "bandwidth" for each format (which would necessarily include resolution, number of planes, size per plane, colour depth, etc.) so you get a nice single numerical value representing a format and then you can just compare the value of A and B.

This seems like a nice idea, I'll just have to look into the specifics. Adjusted argument names for now.

Reply:
Just a thought here so I don't forget; the sort order may also want to cleanly delineate aspect ratio in some way at the top level; i.e. all 16:9 formats separated from all 4:3 formats, etc. for any portrait or square formats that might be in there.

Reply:
Pushed an update to the sort comparator; it now sorts by aspect ratio, then computed bandwidth, then color space, which seems like a better ordering.

Bandwidth is computed a bit lazily, relying partially on a simple bitsPerPixel block with me providing the values. Since this will only used inside this sort comparator and it mostly just wants to be fast, it seemed okay. Not sure what other formats I may want to include values for, though.

As I'm thinking about this more, I'm wondering if, after aspect ratio, we may actually want to sort by (resolution * fps) first, and then bits per pixel second, possibly in reverse order. Reason being, If we're listing all the formats available at a particular FPS and resolution, from there the "best" option (i.e., the one that most users should pick) may be the one with the lowest bandwidth rather than the highest.

Reply:
So the way you'd normally do this is to create a custom category for `AVCaptureDeviceFormat` (say `AVCaptureDeviceFormat+Comparable.m`) and extend the original class with a lazily computed property and add `- (NSComparisonResult)compare:(MyClass *)object` as the comparator.

That way you'd encapsulate the calculation of the comparison value within the format class and also provide the canonical `compare` function implicitly used by Foundation.

Reply:
Switched it around to compute/store a bandwidth scalar and sort using a selector instead of the NSComparator.

Thoughts now:

1. We could just store a generic "sort order" value instead of a 'bandwidth scalar' value. Since the sort order doesn't seem like it should be strictly/only bandwidth-based, and that way we could avoid computing aspect ratio in the selector. If we did that we could maybe even make a sort descriptor and sort solely on that computed property.
2. still same reverse-bandwidth thoughts as above

Also let me know if you see any other issues.

Reply:
Got on a bit of a roll so went ahead and revamped sorting with my idea so it could be tested. Refactored to use sort descriptors on top of the computed properties.

Currently sorts by aspect ratio (rounded to two places) descending, then 'pixels per second' (i.e. resolution * fps) descending, then 'bits per pixel' *ascending*, with the idea that you would present less bandwidth-intense formats first for the same pixel throughput. Couple screenshots beneath; certainly needs more testing.

<img width="350" alt="Screenshot 2024-07-10 at 1 04 07 AM" src="https://github.com/obsproject/obs-studio/assets/6864788/7a3d7fbf-5bfc-444e-9645-e72a8aca1bd8">

<img width="350" alt="Screenshot 2024-07-10 at 1 14 52 AM" src="https://github.com/obsproject/obs-studio/assets/6864788/5fee2a66-d4bc-4612-b79a-ea1bf38c40e8">